source/manual/intro-to-docker-even-more.html.md

```diff
@@ -0,0 +1,398 @@
+
+```
+volumes:
+  - home:/home/build
+  - ~/govuk:/govuk
+```
+
+* Change our `Makefile` to install Ruby and bundler
+
+```
+content-publisher:
+  docker-compose run content-publisher-demo rbenv install -s
+  docker-compose run content-publisher-demo sh -c 'gem install --conservative --no-document bundler -v $$(grep -A1 "BUNDLED WITH" Gemfile.lock | tail -1)'
+  docker-compose run content-publisher-demo bundle install
+  docker-compose run content-publisher-demo bundle exec rake db:setup
+
+```
+
+* Get the tests passing now we're running with rbenv
+
+```shell
+# it can take some time to install Ruby!
+$mac make content-publisher
+
+# run the tests
+$mac docker-compose run content-publisher-demo bundle exec rake
+```
+
+## Step 4: Rails
+
+Now the tests are passing, we're going to shift our focus and look at running Content Publisher as a web app. The aim is to get it running in our browser at `content-publisher-demo.intro-to-docker.gov.uk`.
+
+Our host machine can't communicate with the process in the container. We need to expose the Rails port (3000) on our host machine, and tell Rails to listen for requests on all network interfaces.
+
+* Start with a hacky bunch of command flags
+
+```shell
+$mac docker-compose run -p 80:3000 content-publisher-demo bin/rails s -b 0.0.0.0
+```
+
+> If you see a `Bind for 0.0.0.0:80 failed` error, you need to stop the process on your local machine that's using this port. This is probably GOV.UK Docker, so try doing `govuk-docker down`.
+
+> If you see a `A server is already running` error, you need to add an additional `--restart` flag to tell Rails to cleanup its `tmp/pids` directory, which sometimes fails when containers exit too quickly.
+
+If you visit `localhost` in your browser, you should see some Rails logs in the terminal, and the Content Publisher homepage should (eventually) be visible. Now let's tidy up.
+
+* Move the binding flag for Rails into config
+
+```
+environment:
+  BINDING: 0.0.0.0
+```
+
+* Run `rails s` as the default command
+
+```
+command: bin/rails s
+```
+
+* Re-run with a slightly shorter command
+
+```shell
+$mac docker-compose run -p 80:3000 content-publisher-demo
+```
+
+* Add a temporary entry in `/etc/hosts`
+
+```
+127.0.0.1 content-publisher-demo.intro-to-docker.gov.uk
+```
+
+You should now be able to go to this domain in your browser and see the Content Publisher homepage. After the next section we'll iterate this approach to make it more dynamic and work across multiple apps.
+
+## Step 5: YAML
+
+The `content-publisher-demo` service in our `docker-compose.yml` file is now serving two purposes: running tests and running the app. As we add more environment variables, dependencies, etc., it makes sense to have variants of `content-publisher-demo` that have just the right amount of config for what they need to do.
+
+* Move common config into [an extension][yaml-extension]
+
+```
+# this feature needs a higher version of compose
```

Comment:
Which feature? Aliasing?

Reply:
I've reworded this slightly - it refers to the bullet this is part of. Any better?